include/internal/quic_fc.h

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_QUIC_FC_H
+# define OSSL_QUIC_FC_H
+
+# include <openssl/ssl.h>
+# include "internal/time.h"
+
+/*
+ * TX Flow Controller (TXFC)
+ * =========================
+ *
+ * For discussion, see doc/designs/quic-design/quic-fc.md.
+ */
+typedef struct quic_txfc_st {
+    uint64_t swm, cwm;
+    char     is_stream, has_become_blocked;
+} QUIC_TXFC;
+
+/*
+ * Initialises a TX flow controller. is_stream should be 1 if the TXFC is for
+ * stream-level flow control, and 0 otherwise.
+ */
+int ossl_quic_txfc_init(QUIC_TXFC *txfc, int is_stream);
+
+/*
+ * Bump the credit watermark (CWM) value. This is the 'On TX Window Updated'
+ * operation. This function is a no-op if it has already been called with an
+ * equal or higher CWM value.
+ *
+ * It returns 1 iff the call resulted in the CWM being bumped and 0 if it was
+ * not increased because it has already been called with an equal or higher CWM
+ * value. This is not an error per se but may indicate a local programming error
+ * or a protocol error in a remote peer.
+ */
+int ossl_quic_txfc_bump_cwm(QUIC_TXFC *txfc, uint64_t cwm);
+
+/*
+ * Get the number of bytes by which we are in credit. This is the number of
+ * controlled bytes we are allowed to send. (Thus if this function returns 0, we
+ * are currently blocked.)
+ *
+ * As a convenience, if called on a stream-level TXFC and conn_txfc is non-NULL,
+ * ossl_quic_txfc_get_credit() is called on that TXFC as well, and the lesser of
+ * the two values are returned. conn_txfc must be NULL on a connection-level
+ * TXFC.
+ */
+uint64_t ossl_quic_txfc_get_credit(QUIC_TXFC *txfc, QUIC_TXFC *conn_txfc);
```

Comment:
I think it would be better if the stream FC contained a pointer to its connection's FC.  On both the RX and TX sides.

This avoids callers needing to _know_ the connection between the two and having the pair passed all around the place.